File: Cracking the Coding Interview/Queues- A Tale of Two Stacks/QueuesATaleofTwoStacks.py
# https://www.hackerrank.com/challenges/ctci-queue-using-two-stacks/problem

class MyQueue(object):
    def __init__(self):
        self.stackNewest = []
        self.stackOldest = []
    
    def peek(self):
        self.transfer()
        return self.stackOldest[-1]
        
    def pop(self):
        self.transfer()
        return self.stackOldest.pop()
        
    def put(self, value):
        self.stackNewest.append(value)
        
    def transfer(self):
        if len(self.stackOldest) == 0:
            while len(self.stackNewest) != 0:
                self.stackOldest.append(self.stackNewest.pop())

queue = MyQueue()
t = int(input())
for line in range(t):
    values = map(int, input().split())
    values = list(values)
    if values[0] == 1:
        queue.put(values[1])        
    elif values[0] == 2:
        queue.pop()
    else:
        print(queue.peek())
        

File: Cracking the Coding Interview/Queues- A Tale of Two Stacks/QueuesATaleofTwoStacks.java
// https://www.hackerrank.com/challenges/ctci-queue-using-two-stacks/problem

import java.io.*;
import java.util.*;
import java.text.*;
import java.math.*;
import java.util.regex.*;

public class Solution {
    
    public static class MyQueue<T> {
        Stack<T> stackNewestOnTop = new Stack<T>();
        Stack<T> stackOldestOnTop = new Stack<T>();

        public void enqueue(T value) { // Push onto newest stack
            stackNewestOnTop.push(value);
        }

        public T peek() {
            migrate();
            return stackOldestOnTop.peek();
        }

        public T dequeue() {
            migrate();
            return stackOldestOnTop.pop();
        }
        
        private void migrate() {
            if (stackOldestOnTop.isEmpty()) {
                while(!stackNewestOnTop.isEmpty()) {
                    stackOldestOnTop.push(stackNewestOnTop.pop());
                }
            }
        }
    }

    
    public static void main(String[] args) {
        MyQueue<Integer> queue = new MyQueue<Integer>();
        
        Scanner scan = new Scanner(System.in);
        int n = scan.nextInt();
        
        for (int i = 0; i < n; i++) {
            int operation = scan.nextInt();
            if (operation == 1) { // enqueue
                queue.enqueue(scan.nextInt());
            } else if (operation == 2) { // dequeue
                queue.dequeue();
            } else if (operation == 3) { // print/peek
                System.out.println(queue.peek());
            }
        }
        scan.close();
    }
}
